add option to provide a cipher #59

mdpadberg opened the pull request:
add option to provide a cipher #59

martinvisser commented:
Thanks for the PR, but I have some questions/remarks. I do not see a reason why we need an interface for the encryption service. We could opt for the current `EncryptionService` and if no Cipher is defined it would use the current default Cipher. If the "old" `withEncryption(PublicKey)` is used, it could init the Cipher and basically call `withCipher(Cipher)` method. This would mean there is no need for multiple classes to implement an interface, the `EncryptionService` will always be initialized with a Cipher.

We also might need to clarify that calling both `withEncryption` and `withCipher` will override each other, that's not clear in the docs.

martinvisser reviewed (changes requested):
Thanks for the PR, but I have some questions/remarks. I do not see a reason why we need an interface for the encryption service. We could opt for the current `EncryptionService` and if no Cipher is defined it would use the current default Cipher. If the "old" `withEncryption(PublicKey)` is used, it could init the Cipher and basically call `withCipher(Cipher)` method. This would mean there is no need for multiple classes to implement an interface, the `EncryptionService` will always be initialized with a Cipher.

We also might need to clarify that calling both `withEncryption` and `withCipher` will override each other, that's not clear in the docs.

mdpadberg commented:
> Thanks for the PR, but I have some questions/remarks. I do not see a reason why we need an interface for the encryption service. We could opt for the current `EncryptionService` and if no Cipher is defined it would use the current default Cipher. If the "old" `withEncryption(PublicKey)` is used, it could init the Cipher and basically call `withCipher(Cipher)` method. This would mean there is no need for multiple classes to implement an interface, the `EncryptionService` will always be initialized with a Cipher.
> 
> We also might need to clarify that calling both `withEncryption` and `withCipher` will override each other, that's not clear in the docs.

Updated the javadocs. Also changed the design from the `EncryptionService` interface to the builder, like you suggested

martinvisser commented on src/main/java/io/github/rabobank/shadow_tool/ShadowFlow.java:
```diff
@@ -247,14 +248,38 @@ public ShadowFlowBuilder<T> withExecutorService(final ExecutorService executorSe
         public ShadowFlowBuilder<T> withEncryption(final PublicKey publicKey) {
             try {
-                encryptionService = new EncryptionService(publicKey);
+                final String ALGORITHM = "RSA";
+                final String ALGORITHM_MODE_PADDING = ALGORITHM + "/ECB/OAEPWITHSHA-256ANDMGF1PADDING";
```
Please move these to "constants" at the top of the class (`private static final`)

mdpadberg replied:
Happy to change it if you want. But will it ever be used outside of this method? Because this is pretty specific for the creation of this cipher

martinvisser replied:
Well, it would be (very) slightly better, because they would be static. On the other hand it only happens in the builder. It's fine like this!

martinvisser commented on src/main/java/io/github/rabobank/shadow_tool/EncryptionService.java:
```diff
@@ -3,22 +3,16 @@
     }
 
-    String encrypt(final String value) {
+    public String encrypt(final String value) {
```
I don't see a need for this to be public, package private should be fine.

mdpadberg replied:
My bad, not sure when and why this happened
